### PR TITLE
RATIS-1696. Support linearizable read-only requests in leader

### PR DESCRIPTION
**Pull request by SzyWilliam**
## What changes were proposed in this pull request?
Implement linearizable read-only requests in leader, according to Raft Thesis section 6.4.
1. Obtain `readIndex` by broadcasting heartbeats and receiving majority acks.
2. Wait statemachine to advance at least as far as `readIndex`
3. Query statemachine now and we can get a linearizable result.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/RATIS-1696

Please replace this section with the link to the Apache JIRA)

## How was this patch tested?
New unit tests in ReadOnlyTests, including basic / parallel client / timeout scenarios.

### CODE REVIEW

**Comment by SzyWilliam** [comment]
@szetszwo Hi, last time we have discussions on moving `ReadRequests` to `LeaderStateImpl`. It is inconvenient since we cannot initialize the `appliedIndexConsumer` in `StateMachineUpdater` constructor. Thus, I am thinking about separating `ReadIndexQueue` from `ReadRequests`, since
1. `ReadIndexQueue` should be accessed by both leader and follower peer after we implement **follower linearizable read**.
2. After separating, we can avoid the above inconveniences, placing the `ReadRequests` in `LeaderStateImpl` and `ReadIndexQueue` in `ServerState`.
What do you think?

File: ratis-common/src/main/java/org/apache/ratis/protocol/exceptions/ReadOnlyException.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ratis.protocol.exceptions;
+
+/**
+ * This exception indicates the failure of a read-only request.
+ */
+public class ReadOnlyException extends RaftException {
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply ratis-common/src/main/java/org/apache/ratis/protocol/exceptions/ReadOnlyException.java:23]: done

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java [hunk 2]
```diff
@@ -104,7 +117,7 @@ boolean receive(AppendEntriesReplyProto proto, Predicate<Predicate<RaftPeerId>>
       }
       final RaftProtos.RaftRpcReplyProto rpc = proto.getServerReply();
       final HeartbeatAck reply = replies.get(RaftPeerId.valueOf(rpc.getReplyId()));
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java:119]: done

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java [hunk 3]
```diff
@@ -129,28 +142,75 @@ synchronized AppendEntriesListener add(long commitIndex, Function<Long, AppendEn
...
+          final AppendEntriesListener removed = sorted.remove(entry.getKey());
           Preconditions.assertSame(listener, removed, "AppendEntriesListener");
           ackedCommitIndex.updateToMax(listener.commitIndex, s -> LOG.debug("{}: {}", ReadRequests.this, s));
         }
       }
     }
   }
 
+  static class ReadIndexQueue {
+    private final TimeoutScheduler scheduler = TimeoutScheduler.getInstance();
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java:165]: done

File: ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java [hunk 4]
```diff
@@ -220,6 +224,7 @@ private void reload() throws IOException {
     final long i = snapshot.getIndex();
     snapshotIndex.setUnconditionally(i, infoIndexChange);
     appliedIndex.setUnconditionally(i, infoIndexChange);
+    notifyAppliedIndex(appliedIndex.get());
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java:227]: done

File: ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java [hunk 5]
```diff
@@ -242,6 +247,7 @@ private MemoizedSupplier<List<CompletableFuture<Message>>> applyLog() throws Raf
         }
         final long incremented = appliedIndex.incrementAndGet(debugIndexChange);
         Preconditions.assertTrue(incremented == nextIndex);
+        notifyAppliedIndex(appliedIndex.get());
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java:250]: done

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java [hunk 6]
```diff
@@ -907,6 +910,43 @@ private CompletableFuture<RaftClientReply> staleReadAsync(RaftClientRequest requ
...
+      if (leader == null) {
+        return JavaUtils.completeExceptionally(generateNotLeaderException());
+      }
+      return leader.getReadIndex()
+          .thenCompose(readIndex -> getReadRequests().waitToAdvance(readIndex))
+          .thenCompose(readIndex -> queryStateMachine(request))
+          .exceptionally(e -> readOnlyException2Reply(request, e));
+    }
+    // default
+    return queryStateMachine(request);
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:936]: done

File: ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java [hunk 7]
```diff
@@ -129,28 +142,75 @@ synchronized AppendEntriesListener add(long commitIndex, Function<Long, AppendEn
     synchronized void onAppendEntriesReply(AppendEntriesReplyProto reply,
                                            Predicate<Predicate<RaftPeerId>> hasMajority) {
       final long callId = reply.getServerReply().getCallId();
-      for (;;) {
-        final Map.Entry<Long, AppendEntriesListener> first = sorted.firstEntry();
-        if (first == null || first.getKey() > callId) {
+      for (Map.Entry<Long, AppendEntriesListener> entry : sorted.entrySet()) {
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java:145]: I'll use Iterator here to avoid `ConcurrentModificationException`.

**Comment by SzyWilliam** [comment]
@szetszwo Thanks for the reviews. I've made changes on code, please take a look again~

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java [hunk 8]
```diff
@@ -1110,8 +1109,9 @@ CompletableFuture<Long> getReadIndex() {
   }
 
   @Override
-  public void onAppendEntriesReply(FollowerInfo follower, RaftProtos.AppendEntriesReplyProto reply) {
-    server.getState().getReadRequests().onAppendEntriesReply(reply, this::hasMajority);
+  public void onAppendEntriesReply(LogAppender appender, FollowerInfo follower,
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java:1112]: Remove `FollowerInfo` since it is unused.

**Review by szetszwo — COMMENTED** [review]
@SzyWilliam , the change looks good.  Just a minor comment inlined.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java [hunk 9]
```diff
@@ -1110,8 +1109,9 @@ CompletableFuture<Long> getReadIndex() {
...
+                                   RaftProtos.AppendEntriesReplyProto reply) {
+    server.getState().getReadRequests().onAppendEntriesReply(appender, reply, this::hasMajority);
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java:1114]: One more comment: Use `server.getReadRequests()`  instead in line 1114 (this line) and line 1090.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java (hunk 8 above)
Comment (SzyWilliam) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java]: done

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java (hunk 9 above)
Comment (SzyWilliam) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java]: done

**Comment by SzyWilliam** [comment]
@szetszwo Thanks for the reviews again. Since it is already a complex change, I'll separate ReadRequests and ReadIndexQueue in the next PR.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**